File: mb_tagger/deezer.py
"""Deezer album genre + per-track explicit lookups (public API, disk-cached)."""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from difflib import SequenceMatcher
from urllib.parse import parse_qs, urlparse

import requests
from requests.adapters import HTTPAdapter

from mb_tagger.constants import DEEZER_BASE, DEEZER_HEADERS
from mb_tagger.genres import pretty_genres
from mb_tagger.mb_cache import cache_get, cache_set
from mb_tagger.pacer import Pacer
from mb_tagger.titles import strip_mark

_dz_http_requests = 0
_dz_cache_hits = 0
_MIN_INTERVAL = 0.12
_SKIP_GENRES = {"unclassified", "unknown", "other", "none"}
_PAREN_TAIL = re.compile(r"\s*\([^)]*\)\s*$")
_album_memo: dict[tuple[str, str], "DeezerAlbumMatch"] = {}
_album_id_memo: dict[int, "DeezerAlbumMatch"] = {}
_artist_search_memo: dict[str, "DeezerArtistInfo | None"] = {}
_dz_session: requests.Session | None = None
_pacer = Pacer(_MIN_INTERVAL)


def dz_stats() -> dict[str, int]:
    return {"deezer_http": _dz_http_requests, "deezer_cache_hits": _dz_cache_hits}


def reset_dz_stats() -> None:
    global _dz_http_requests, _dz_cache_hits
    _dz_http_requests = 0
    _dz_cache_hits = 0
    _album_memo.clear()
    _album_id_memo.clear()
    _artist_search_memo.clear()


def _session() -> requests.Session:
    global _dz_session
    if _dz_session is None:
        _dz_session = requests.Session()
        _dz_session.headers.update(DEEZER_HEADERS)
        adapter = HTTPAdapter(pool_connections=8, pool_maxsize=8)
        _dz_session.mount("http://", adapter)
        _dz_session.mount("https://", adapter)
    return _dz_session


@dataclass
class DeezerTrack:
    title: str
    explicit: bool | None = None
    artists: list[str] = field(default_factory=list)


@dataclass
class DeezerArtistInfo:
    name: str
    artist_id: int | None = None
    picture: str = ""


@dataclass
class DeezerAlbumMatch:
    genres: list[str] = field(default_factory=list)
    source: str = "no-match"
    album_id: int | None = None
    title: str = ""
    album_artist: str = ""
    artists: list[str] = field(default_factory=list)
    artist_infos: list[DeezerArtistInfo] = field(default_factory=list)
    tracks: list[DeezerTrack] = field(default_factory=list)
    explicit: bool | None = None


def _norm(text: str) -> str:
    s = strip_mark(text or "").lower()
    s = re.sub(r"[^\w\s&]+", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def _album_variants(album: str) -> list[str]:
    variants: list[str] = []
    current = (album or "").strip()
    while current:
        if current not in variants:
            variants.append(current)
        nxt = _PAREN_TAIL.sub("", current).strip()
        if nxt == current:
            break
        current = nxt
    return variants


def _quote(text: str) -> str:
    return (text or "").replace('"', "").strip()


def dz_get(path: str, params: dict | None = None) -> dict:
    """Deezer GET with disk cache, pacing, and retries on 429/5xx."""
    global _dz_http_requests, _dz_cache_hits
    params = dict(params or {})
    cache_path = f"deezer/{path.lstrip('/')}"
    cached = cache_get(cache_path, params)
    if cached is not None:
        _dz_cache_hits += 1
        return cached

    url = f"{DEEZER_BASE}/{path.lstrip('/')}"
    last_err: Exception | None = None
    for attempt in range(5):
        _pacer.wait()
        try:
            res = _session().get(url, params=params or None, timeout=30)
            if res.status_code in {429, 500, 502, 503, 504}:
                time.sleep(1.5 * (attempt + 1))
                last_err = requests.HTTPError(f"{res.status_code} {res.reason}", response=res)
                continue
            res.raise_for_status()
            data = res.json()
            if not isinstance(data, dict):
                return {}
            if data.get("error"):
                return data
            _dz_http_requests += 1
            cache_set(cache_path, params, data)
            return data
        except (requests.RequestException, ValueError) as exc:
            last_err = exc
            time.sleep(1.5 * (attempt + 1))
    if last_err:
        raise last_err
    return {}


def download_image(url: str) -> tuple[bytes, str]:
    """Fetch a Deezer CDN image. Returns (bytes, content-type)."""
    res = _session().get(url, timeout=30)
    res.raise_for_status()
    mime = (res.headers.get("Content-Type") or "image/jpeg").split(";")[0].strip().lower()
    if mime in {"image/jpg", "jpg", "jpeg"} or not mime.startswith("image/"):
        mime = "image/jpeg"
    return res.content, mime


def _picture_url(payload: dict) -> str:
    return str(
        payload.get("picture_xl")
        or payload.get("picture_big")
        or payload.get("picture")
        or ""
    ).strip()


def _artist_infos(payload: dict) -> list[DeezerArtistInfo]:
    infos: list[DeezerArtistInfo] = []
    seen: set[str] = set()

    def add(raw: object) -> None:
        if not isinstance(raw, dict):
            return
        name = str(raw.get("name") or "").strip()
        if not name:
            return
        key = name.lower()
        if key in seen:
            return
        seen.add(key)
        artist_id = raw.get("id")
        picture = _picture_url(raw)
        infos.append(
            DeezerArtistInfo(
                name=name,
                artist_id=int(artist_id) if artist_id else None,
                picture=picture,
            )
        )

    add(payload.get("artist"))
    for person in payload.get("contributors") or []:
        add(person)
    return infos


def _genres_from_album_payload(payload: dict) -> list[str]:
    names: list[str] = []
    for g in ((payload.get("genres") or {}).get("data") or []):
        name = str(g.get("name") or "").strip()
        if not name or _norm(name) in _SKIP_GENRES:
            continue
        names.append(name)
    return pretty_genres(names, max_genres=3)


def _explicit_from_payload(payload: dict, *, album: bool = False) -> bool | None:
    """Deezer explicit_content_lyrics: 0 clean, 1 explicit, 2 unknown, 3 edited, 4 partial (album)."""
    code = payload.get("explicit_content_lyrics")
    if code == 1:
        return True
    if album and code == 4:
        return True
    if code in (0, 3):
        return False
    if code == 2:
        return None
    flag = payload.get("explicit_lyrics")
    if flag is True:
        return True
    if flag is False:
        return False
    return None


def _explicit_from_track(payload: dict) -> bool | None:
    """Track-level only. Album code 4 (partially explicit) is not inherited onto tracks."""
    return _explicit_from_payload(payload, album=False)


def _explicit_rank(payload: dict | DeezerTrack, *, album: bool = False) -> int:
    """Higher is better: explicit original > unknown > clean/edited."""
    if isinstance(payload, DeezerTrack):
        flag = payload.explicit
    else:
        flag = _explicit_from_payload(payload, album=album)
    if flag is True:
        return 2
    if flag is False:
        return 0
    return 1


def _artist_ok(got: str, want: str) -> bool:
    if not want:
        return True
    if not got:
        return False
    return got == want or want in got or got in want


def _pick_album(results: list[dict], artist: str, album: str) -> dict | None:
    want_artist = _norm(artist)
    want_album = _norm(album)
    best: dict | None = None
    best_key = (-1.0, -1)
    for item in results:
        got_artist = _norm((item.get("artist") or {}).get("name") or "")
        if not _artist_ok(got_artist, want_artist):
            continue
        got_title = _norm(item.get("title") or "")
        if not got_title:
            continue
        if got_title == want_album:
            score = 1.0
        else:
            score = SequenceMatcher(None, got_title, want_album).ratio()
            if want_album in got_title or got_title in want_album:
                score = max(score, 0.82)
        key = (score, _explicit_rank(item, album=True))
        if key > best_key:
            best, best_key = item, key
    if best is not None and best_key[0] >= 0.72:
        return best
    return None


def _album_tracks(album_id: int) -> list[DeezerTrack]:
    items: list[DeezerTrack] = []
    path = f"album/{album_id}/tracks"
    params: dict[str, str] | None = {"limit": "100"}
    while path:
        payload = dz_get(path, params)
        for raw in payload.get("data") or []:
            title = str(raw.get("title") or "").strip()
            if not title:
                continue
            items.append(
                DeezerTrack(
                    title=title,
                    explicit=_explicit_from_track(raw),
                    artists=[info.name for info in _artist_infos(raw)],
                )
            )
        nxt = payload.get("next")
        if not nxt:
            break
        parsed = urlparse(str(nxt))
        path = parsed.path.lstrip("/")
        if path.startswith("2.0/"):
            path = path[4:]
        params = {k: v[0] for k, v in parse_qs(parsed.query).items()}
    return items


def _album_by_id(album_id: int) -> DeezerAlbumMatch:
    cached = _album_id_memo.get(album_id)
    if cached is not None:
        return cached
    payload = dz_get(f"album/{album_id}")
    if payload.get("error") or not payload.get("id"):
        match = DeezerAlbumMatch(source="no-match")
        _album_id_memo[album_id] = match
        return match
    genres = _genres_from_album_payload(payload)
    artist_infos = _artist_infos(payload)
    artists = [info.name for info in artist_infos]
    tracks = _album_tracks(album_id)
    source = f"album:{album_id}"
    if not genres and not tracks:
        source = "no-genre"
    match = DeezerAlbumMatch(
        genres=genres,
        source=source,
        album_id=int(payload.get("id") or album_id),
        title=str(payload.get("title") or ""),
        album_artist=artists[0] if artists else "",
        artists=artists,
        artist_infos=artist_infos,
        tracks=tracks,
        explicit=_explicit_from_payload(payload, album=True),
    )
    _album_id_memo[album_id] = match
    return match


def _search_album(artist: str, album: str) -> DeezerAlbumMatch:
    q = f'artist:"{_quote(artist)}" album:"{_quote(album)}"'
    payload = dz_get("search/album", {"q": q, "limit": 25})
    hit = _pick_album(payload.get("data") or [], artist, album)
    if not hit:
        return DeezerAlbumMatch(source="no-match")
    return _album_by_id(int(hit["id"]))


def _search_track_album(artist: str, title: str) -> DeezerAlbumMatch:
    q = f'artist:"{_quote(artist)}" track:"{_quote(strip_mark(title))}"'
    payload = dz_get("search/track", {"q": q, "limit": 15})
    want_artist = _norm(artist)
    want_title = _norm(title)
    best: dict | None = None
    best_key = (-1.0, -1)
    for item in payload.get("data") or []:
        got_artist = _norm((item.get("artist") or {}).get("name") or "")
        got_title = _norm(item.get("title") or "")
        if not _artist_ok(got_artist, want_artist):
            continue
        if not got_title:
            continue
        if got_title == want_title:
            score = 1.0
        else:
            score = SequenceMatcher(None, got_title, want_title).ratio()
            if want_title in got_title or got_title in want_title:
                score = max(score, 0.84)
        if score < 0.72:
            continue
        key = (score, _explicit_rank(item, album=False))
        if key > best_key:
            best, best_key = item, key
    album_id = (best.get("album") or {}).get("id") if best else None
    if album_id:
        return _album_by_id(int(album_id))
    return DeezerAlbumMatch(source="no-match")


def match_deezer_track(title: str, tracks: list[DeezerTrack]) -> DeezerTrack | None:
    want = _norm(title)
    if not want or not tracks:
        return None
    best: DeezerTrack | None = None
    best_key = (-1.0, -1)
    for track in tracks:
        got = _norm(track.title)
        if not got:
            continue
        if got == want:
            score = 1.0
        else:
            score = SequenceMatcher(None, got, want).ratio()
            if want in got or got in want:
                score = max(score, 0.84)
        if score < 0.72:
            continue
        key = (score, _explicit_rank(track))
        if key > best_key:
            best, best_key = track, key
    return best


def search_artist(name: str) -> DeezerArtistInfo | None:
    """Best-name Deezer artist hit, used to fill profile photos."""
    want = _norm(name)
    if not want:
        return None
    if want in _artist_search_memo:
        return _artist_search_memo[want]
    payload = dz_get("search/artist", {"q": name, "limit": 8})
    best: DeezerArtistInfo | None = None
    best_score = 0.0
    for raw in payload.get("data") or []:
        if not isinstance(raw, dict):
            continue
        got = str(raw.get("name") or "").strip()
        if not got:
            continue
        got_n = _norm(got)
        score = 1.0 if got_n == want else SequenceMatcher(None, got_n, want).ratio()
        if score < 0.86:
            continue
        if score > best_score:
            best_score = score
            artist_id = raw.get("id")
            best = DeezerArtistInfo(
                name=got,
                artist_id=int(artist_id) if artist_id else None,
                picture=_picture_url(raw),
            )
    if best and not best.picture and best.artist_id:
        detail = dz_get(f"artist/{best.artist_id}")
        if not detail.get("error"):
            best.picture = _picture_url(detail)
    _artist_search_memo[want] = best
    return best


def lookup_album(
    artist: str,
    album: str,
    *,
    sample_title: str = "",
) -> DeezerAlbumMatch:
    """Resolve Deezer album genres + track list. Memoized per (artist, album)."""
    key = (_norm(artist), _norm(album))
    if key in _album_memo:
        return _album_memo[key]

    match = DeezerAlbumMatch(source="no-match")
    for variant in _album_variants(album):
        match = _search_album(artist, variant)
        if match.album_id:
            break

    if not match.album_id and sample_title:
        track_match = _search_track_album(artist, sample_title)
        if track_match.album_id:
            match = track_match

    _album_memo[key] = match
    return match

File: internal/deezer/deezer.go
package deezer

import (
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/TidBits16/peanut-butter-jelly/internal/cache"
	"github.com/TidBits16/peanut-butter-jelly/internal/config"
	"github.com/TidBits16/peanut-butter-jelly/internal/genres"
	"github.com/TidBits16/peanut-butter-jelly/internal/httpx"
	"github.com/TidBits16/peanut-butter-jelly/internal/titles"
)

type Track struct {
	Title    string
	Explicit *bool
	Artists  []string
}

type ArtistInfo struct {
	Name     string
	ArtistID int
	Picture  string
}

type AlbumMatch struct {
	Genres      []string
	Source      string
	AlbumID     int
	Title       string
	AlbumArtist string
	Artists     []string
	ArtistInfos []ArtistInfo
	Tracks      []Track
	Explicit    *bool
}

type Client struct {
	http *httpx.Client
	mu   sync.Mutex
	byKA map[[2]string]AlbumMatch
	byID map[int]AlbumMatch
	arts map[string]*ArtistInfo
}

func New(store *cache.Store) *Client {
	return &Client{
		http: httpx.New(store, map[string]string{
			"User-Agent": "peanut-butter-jelly/" + config.Version + " (deezer; jellyfin)",
			"Accept":     "application/json",
		}, 120*time.Millisecond),
		byKA: map[[2]string]AlbumMatch{},
		byID: map[int]AlbumMatch{},
		arts: map[string]*ArtistInfo{},
	}
}

func (c *Client) Stats() (httpN, hits int) { return c.http.HTTPN, c.http.Hits }

func quote(s string) string { return strings.ReplaceAll(strings.TrimSpace(s), `"`, "") }

func albumVariants(album string) []string {
	var out []string
	cur := strings.TrimSpace(album)
	for cur != "" {
		dup := false
		for _, x := range out {
			if x == cur {
				dup = true
				break
			}
		}
		if !dup {
			out = append(out, cur)
		}
		nxt := strings.TrimSpace(parenTail(cur))
		if nxt == cur {
			break
		}
		cur = nxt
	}
	return out
}

func parenTail(s string) string {
	if i := strings.LastIndex(s, "("); i >= 0 && strings.HasSuffix(s, ")") {
		return strings.TrimSpace(s[:i])
	}
	return s
}

func artistOK(got, want string) bool {
	if want == "" {
		return true
	}
	if got == "" {
		return false
	}
	return got == want || strings.Contains(got, want) || strings.Contains(want, got)
}

func ratio(a, b string) float64 {
	if a == b {
		return 1
	}
	// cheap similarity: longest common subsequence-ish via sequence matcher lite
	return seqRatio(a, b)
}

func seqRatio(a, b string) float64 {
	if a == "" || b == "" {
		return 0
	}
	// SequenceMatcher-ish: 2*matches/len
	matches := 0
	bi := 0
	used := make([]bool, len(b))
	for i := 0; i < len(a); i++ {
		for j := bi; j < len(b); j++ {
			if !used[j] && a[i] == b[j] {
				used[j] = true
				matches++
				bi = j + 1
				break
			}
		}
	}
	return 2 * float64(matches) / float64(len(a)+len(b))
}

func explicitFrom(payload map[string]any, album bool) *bool {
	code, has := payload["explicit_content_lyrics"]
	if has && code != nil {
		n := int(asFloat(code))
		if n == 1 || (album && n == 4) {
			t := true
			return &t
		}
		if n == 0 || n == 3 {
			f := false
			return &f
		}
		if n == 2 {
			return nil
		}
	}
	if v, ok := payload["explicit_lyrics"].(bool); ok {
		return &v
	}
	return nil
}

func asFloat(v any) float64 {
	switch t := v.(type) {
	case float64:
		return t
	default:
		n, _ := strconv.ParseFloat(httpx.AsString(v), 64)
		return n
	}
}

func explicitRank(flag *bool) int {
	if flag == nil {
		return 1
	}
	if *flag {
		return 2
	}
	return 0
}

func pictureURL(payload map[string]any) string {
	for _, k := range []string{"picture_xl", "picture_big", "picture"} {
		s := strings.TrimSpace(httpx.AsString(payload[k]))
		if s != "" {
			return s
		}
	}
	return ""
}

func artistInfos(payload map[string]any) []ArtistInfo {
	var infos []ArtistInfo
	seen := map[string]struct{}{}
	add := func(raw any) {
		m := httpx.AsMap(raw)
		name := strings.TrimSpace(httpx.AsString(m["name"]))
		if name == "" {
			return
		}
		key := strings.ToLower(name)
		if _, ok := seen[key]; ok {
			return
		}
		seen[key] = struct{}{}
		id := int(asFloat(m["id"]))
		infos = append(infos, ArtistInfo{Name: name, ArtistID: id, Picture: pictureURL(m)})
	}
	add(payload["artist"])
	for _, p := range httpx.AsList(payload["contributors"]) {
		add(p)
	}
	return infos
}

func (c *Client) get(path string, params map[string]string) (map[string]any, error) {
	return c.http.GetJSON("deezer/"+strings.TrimLeft(path, "/"), config.DeezerBase+"/"+strings.TrimLeft(path, "/"), params, 0)
}

func trackFrom(raw map[string]any) (Track, bool) {
	title := strings.TrimSpace(httpx.AsString(raw["title"]))
	if title == "" {
		return Track{}, false
	}
	var arts []string
	for _, inf := range artistInfos(raw) {
		arts = append(arts, inf.Name)
	}
	return Track{Title: title, Explicit: explicitFrom(raw, false), Artists: arts}, true
}

func (c *Client) albumTracks(id int, embedded []any, nb any) []Track {
	var items []Track
	seen := map[string]struct{}{}
	for _, raw := range embedded {
		t, ok := trackFrom(httpx.AsMap(raw))
		if !ok {
			continue
		}
		items = append(items, t)
		seen[strings.ToLower(t.Title)] = struct{}{}
	}
	expected := 0
	if nb != nil {
		expected = int(asFloat(nb))
	}
	if expected > 0 && len(items) >= expected {
		return items
	}
	path := "album/" + strconv.Itoa(id) + "/tracks"
	params := map[string]string{"limit": "100"}
	for path != "" {
		payload, err := c.get(path, params)
		if err != nil {
			break
		}
		for _, raw := range httpx.AsList(payload["data"]) {
			t, ok := trackFrom(httpx.AsMap(raw))
			if !ok {
				continue
			}
			k := strings.ToLower(t.Title)
			if _, ok := seen[k]; ok {
				continue
			}
			seen[k] = struct{}{}
			items = append(items, t)
		}
		nxt := strings.TrimSpace(httpx.AsString(payload["next"]))
		if nxt == "" {
			break
		}
		u, err := url.Parse(nxt)
		if err != nil {
			break
		}
		path = strings.TrimLeft(u.Path, "/")
		if strings.HasPrefix(path, "2.0/") {
			path = path[4:]
		}
		params = map[string]string{}
		for k, vs := range u.Query() {
			if len(vs) > 0 {
				params[k] = vs[0]
			}
		}
	}
	return items
}

func genresFrom(payload map[string]any) []string {
	var names []string
	skip := map[string]struct{}{"unclassified": {}, "unknown": {}, "other": {}, "none": {}}
	g := httpx.AsMap(payload["genres"])
	for _, raw := range httpx.AsList(g["data"]) {
		name := strings.TrimSpace(httpx.AsString(httpx.AsMap(raw)["name"]))
		if name == "" {
			continue
		}
		if _, ok := skip[titles.Norm(name)]; ok {
			continue
		}
		names = append(names, name)
	}
	return genres.PrettyList(names, 3)
}

func (c *Client) albumByID(id int) AlbumMatch {
	c.mu.Lock()
	if m, ok := c.byID[id]; ok {
		c.mu.Unlock()
		return m
	}
	c.mu.Unlock()
	payload, err := c.get("album/"+strconv.Itoa(id), nil)
	if err != nil || payload["error"] != nil || payload["id"] == nil {
		m := AlbumMatch{Source: "no-match"}
		c.mu.Lock()
		c.byID[id] = m
		c.mu.Unlock()
		return m
	}
	infos := artistInfos(payload)
	var arts []string
	for _, inf := range infos {
		arts = append(arts, inf.Name)
	}
	tracks := c.albumTracks(id, httpx.AsList(httpx.AsMap(payload["tracks"])["data"]), payload["nb_tracks"])
	src := "album:" + strconv.Itoa(id)
	gs := genresFrom(payload)
	if len(gs) == 0 && len(tracks) == 0 {
		src = "no-genre"
	}
	aa := ""
	if len(arts) > 0 {
		aa = arts[0]
	}
	m := AlbumMatch{
		Genres: gs, Source: src, AlbumID: int(asFloat(payload["id"])),
		Title: httpx.AsString(payload["title"]), AlbumArtist: aa,
		Artists: arts, ArtistInfos: infos, Tracks: tracks,
		Explicit: explicitFrom(payload, true),
	}
	c.mu.Lock()
	c.byID[id] = m
	c.mu.Unlock()
	return m
}

func (c *Client) pickAlbum(results []any, artist, album string) map[string]any {
	wantA, wantB := titles.Norm(artist), titles.Norm(album)
	var best map[string]any
	bestScore, bestRank := -1.0, -1
	for _, raw := range results {
		item := httpx.AsMap(raw)
		gotA := titles.Norm(httpx.AsString(httpx.AsMap(item["artist"])["name"]))
		if !artistOK(gotA, wantA) {
			continue
		}
		gotT := titles.Norm(httpx.AsString(item["title"]))
		if gotT == "" {
			continue
		}
		score := ratio(gotT, wantB)
		if gotT == wantB {
			score = 1
		} else if strings.Contains(gotT, wantB) || strings.Contains(wantB, gotT) {
			if score < 0.82 {
				score = 0.82
			}
		}
		rank := explicitRank(explicitFrom(item, true))
		if score > bestScore || (score == bestScore && rank > bestRank) {
			best, bestScore, bestRank = item, score, rank
		}
	}
	if best != nil && bestScore >= 0.72 {
		return best
	}
	return nil
}

func (c *Client) searchAlbum(artist, album string) AlbumMatch {
	q := `artist:"` + quote(artist) + `" album:"` + quote(album) + `"`
	payload, err := c.get("search/album", map[string]string{"q": q, "limit": "25"})
	if err != nil {
		return AlbumMatch{Source: "no-match"}
	}
	hit := c.pickAlbum(httpx.AsList(payload["data"]), artist, album)
	if hit == nil {
		return AlbumMatch{Source: "no-match"}
	}
	return c.albumByID(int(asFloat(hit["id"])))
}

func (c *Client) searchTrackAlbum(artist, title string) AlbumMatch {
	q := `artist:"` + quote(artist) + `" track:"` + quote(titles.StripMark(title)) + `"`
	payload, err := c.get("search/track", map[string]string{"q": q, "limit": "15"})
	if err != nil {
		return AlbumMatch{Source: "no-match"}
	}
	wantA, wantT := titles.Norm(artist), titles.Norm(title)
	var best map[string]any
	bestScore, bestRank := -1.0, -1
	for _, raw := range httpx.AsList(payload["data"]) {
		item := httpx.AsMap(raw)
		gotA := titles.Norm(httpx.AsString(httpx.AsMap(item["artist"])["name"]))
		gotT := titles.Norm(httpx.AsString(item["title"]))
		if !artistOK(gotA, wantA) || gotT == "" {
			continue
		}
		score := ratio(gotT, wantT)
		if gotT == wantT {
			score = 1
		} else if strings.Contains(gotT, wantT) || strings.Contains(wantT, gotT) {
			if score < 0.84 {
				score = 0.84
			}
		}
		if score < 0.72 {
			continue
		}
		rank := explicitRank(explicitFrom(item, false))
		if score > bestScore || (score == bestScore && rank > bestRank) {
			best, bestScore, bestRank = item, score, rank
		}
	}
	if best == nil {
		return AlbumMatch{Source: "no-match"}
	}
	id := asFloat(httpx.AsMap(best["album"])["id"])
	if id == 0 {
		return AlbumMatch{Source: "no-match"}
	}
	return c.albumByID(int(id))
}

func MatchTrack(title string, tracks []Track) *Track {
	want := titles.Norm(title)
	if want == "" || len(tracks) == 0 {
		return nil
	}
	var best *Track
	bestScore, bestRank := -1.0, -1
	for i := range tracks {
		t := &tracks[i]
		got := titles.Norm(t.Title)
		if got == "" {
			continue
		}
		score := ratio(got, want)
		if got == want {
			score = 1
		} else if strings.Contains(got, want) || strings.Contains(want, got) {
			if score < 0.84 {
				score = 0.84
			}
		}
		if score < 0.72 {
			continue
		}
		rank := explicitRank(t.Explicit)
		if score > bestScore || (score == bestScore && rank > bestRank) {
			best, bestScore, bestRank = t, score, rank
		}
	}
	return best
}

func (c *Client) SearchArtist(name string) *ArtistInfo {
	want := titles.Norm(name)
	if want == "" {
		return nil
	}
	c.mu.Lock()
	if v, ok := c.arts[want]; ok {
		c.mu.Unlock()
		return v
	}
	c.mu.Unlock()
	payload, err := c.get("search/artist", map[string]string{"q": name, "limit": "8"})
	if err != nil {
		return nil
	}
	var best *ArtistInfo
	bestScore := 0.0
	for _, raw := range httpx.AsList(payload["data"]) {
		m := httpx.AsMap(raw)
		got := strings.TrimSpace(httpx.AsString(m["name"]))
		if got == "" {
			continue
		}
		gotN := titles.Norm(got)
		score := ratio(gotN, want)
		if gotN == want {
			score = 1
		}
		if score < 0.86 {
			continue
		}
		if score > bestScore {
			bestScore = score
			best = &ArtistInfo{Name: got, ArtistID: int(asFloat(m["id"])), Picture: pictureURL(m)}
		}
	}
	if best != nil && best.Picture == "" && best.ArtistID != 0 {
		detail, err := c.get("artist/"+strconv.Itoa(best.ArtistID), nil)
		if err == nil && detail["error"] == nil {
			best.Picture = pictureURL(detail)
		}
	}
	c.mu.Lock()
	c.arts[want] = best
	c.mu.Unlock()
	return best
}

func (c *Client) LookupAlbum(artist, album, sampleTitle string) AlbumMatch {
	key := [2]string{titles.Norm(artist), titles.Norm(album)}
	c.mu.Lock()
	if m, ok := c.byKA[key]; ok {
		c.mu.Unlock()
		return m
	}
	c.mu.Unlock()
	var disk map[string]any
	if c.http.Cache != nil && c.http.Cache.Get("deezer/album-match/v1", map[string]string{"artist": key[0], "album": key[1]}, &disk, 0) {
		m := matchFromCache(disk)
		c.mu.Lock()
		c.http.Hits++
		c.byKA[key] = m
		if m.AlbumID != 0 {
			c.byID[m.AlbumID] = m
		}
		c.mu.Unlock()
		return m
	}
	m := AlbumMatch{Source: "no-match"}
	for _, variant := range albumVariants(album) {
		m = c.searchAlbum(artist, variant)
		if m.AlbumID != 0 {
			break
		}
	}
	if m.AlbumID == 0 && sampleTitle != "" {
		tm := c.searchTrackAlbum(artist, sampleTitle)
		if tm.AlbumID != 0 {
			m = tm
		}
	}
	if c.http.Cache != nil {
		c.http.Cache.Set("deezer/album-match/v1", map[string]string{"artist": key[0], "album": key[1]}, matchToCache(m))
	}
	c.mu.Lock()
	c.byKA[key] = m
	if m.AlbumID != 0 {
		c.byID[m.AlbumID] = m
	}
	c.mu.Unlock()
	return m
}

func (c *Client) DownloadImage(url string) ([]byte, string, error) {
	b, mime, err := c.http.GetBytes(url)
	if err != nil {
		return nil, "", err
	}
	mime = strings.ToLower(strings.Split(mime, ";")[0])
	if mime == "image/jpg" || mime == "jpg" || mime == "jpeg" || !strings.HasPrefix(mime, "image/") {
		mime = "image/jpeg"
	}
	return b, mime, nil
}

func boolPtrFrom(v any) *bool {
	if v == nil {
		return nil
	}
	if b, ok := v.(bool); ok {
		return &b
	}
	return nil
}

func matchFromCache(raw map[string]any) AlbumMatch {
	m := AlbumMatch{
		Source:      httpx.AsString(raw["source"]),
		Title:       httpx.AsString(raw["title"]),
		AlbumArtist: httpx.AsString(raw["album_artist"]),
		Explicit:    boolPtrFrom(raw["explicit"]),
	}
	if id := asFloat(raw["album_id"]); id != 0 {
		m.AlbumID = int(id)
	}
	for _, g := range httpx.AsList(raw["genres"]) {
		m.Genres = append(m.Genres, httpx.AsString(g))
	}
	for _, a := range httpx.AsList(raw["artists"]) {
		m.Artists = append(m.Artists, httpx.AsString(a))
	}
	for _, inf := range httpx.AsList(raw["artist_infos"]) {
		im := httpx.AsMap(inf)
		m.ArtistInfos = append(m.ArtistInfos, ArtistInfo{
			Name: httpx.AsString(im["name"]), ArtistID: int(asFloat(im["artist_id"])), Picture: httpx.AsString(im["picture"]),
		})
	}
	for _, t := range httpx.AsList(raw["tracks"]) {
		tm := httpx.AsMap(t)
		var arts []string
		for _, a := range httpx.AsList(tm["artists"]) {
			arts = append(arts, httpx.AsString(a))
		}
		m.Tracks = append(m.Tracks, Track{Title: httpx.AsString(tm["title"]), Explicit: boolPtrFrom(tm["explicit"]), Artists: arts})
	}
	return m
}

func matchToCache(m AlbumMatch) map[string]any {
	infos := []any{}
	for _, i := range m.ArtistInfos {
		infos = append(infos, map[string]any{"name": i.Name, "artist_id": i.ArtistID, "picture": i.Picture})
	}
	tracks := []any{}
	for _, t := range m.Tracks {
		var ex any
		if t.Explicit != nil {
			ex = *t.Explicit
		}
		tracks = append(tracks, map[string]any{"title": t.Title, "explicit": ex, "artists": t.Artists})
	}
	var ex any
	if m.Explicit != nil {
		ex = *m.Explicit
	}
	return map[string]any{
		"genres": m.Genres, "source": m.Source, "album_id": m.AlbumID, "title": m.Title,
		"album_artist": m.AlbumArtist, "artists": m.Artists, "explicit": ex,
		"artist_infos": infos, "tracks": tracks,
	}
}
